File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagna

const NoodleGramsPerLayer int = 50
const SauceLitersPerLayer float64 = 0.2

// TODO: define the 'PreparationTime()' function
func PreparationTime(layers []string, prepTimePerLayer int) int {
	if prepTimePerLayer == 0 {
		prepTimePerLayer = 2
	}

	return len(layers) * prepTimePerLayer
}

// TODO: define the 'Quantities()' function
func Quantities(layers []string) (int, float64) {
	var gramsOfNoodles int = 0
	var litersOfSauce float64 = 0

	for _, layer := range(layers) {
		switch layer {
		case "sauce":
			litersOfSauce += SauceLitersPerLayer
		case "noodles":
			gramsOfNoodles += NoodleGramsPerLayer
		}
	}

	return gramsOfNoodles, litersOfSauce
}

// TODO: define the 'AddSecretIngredient()' function
func AddSecretIngredient(friendsList []string, myList []string) {
	myList[len(myList) - 1] = friendsList[len(friendsList) - 1]
}

// TODO: define the 'ScaleRecipe()' function
func ScaleRecipe(quantities []float64, portions int) []float64 {
	var scaleFactor float64 = float64(portions) / 2.0
	var scaledQuantities = make([]float64, len(quantities))

	for i,_ := range(quantities) {
		scaledQuantities[i] = scaleFactor * quantities[i]
	}

	return scaledQuantities
}

// Your first steps could be to read through the tasks, and create
// these functions with their correct parameter lists and return types.
// The function body only needs to contain `panic("")`.
// 
// This will make the tests compile, but they will fail.
// You can then implement the function logic one by one and see
// an increasing number of tests passing as you implement more 
// functionality.

File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export const GRAMS_NOODLES_PER_LAYER = 50
export const LITERS_SAUCE_PER_LAYER = 0.2

export function cookingStatus(remainingTime) {
  switch (remainingTime) {
    case undefined:
      return 'You forgot to set the timer.'
    case 0:
      return 'Lasagna is done.'
    default:
      return 'Not done, please wait.'
  }
}

export function preparationTime(layers, prepTimePerLayer = 2) {
  return prepTimePerLayer * layers.length
}

export function quantities(layers) {
  let obj = {
    noodles: 0,
    sauce: 0,
  }

  layers.forEach((layer) => {
    switch (layer) {
      case "noodles":
        obj.noodles += GRAMS_NOODLES_PER_LAYER
        break
      case "sauce":
        obj.sauce += LITERS_SAUCE_PER_LAYER
    }
  })

  return obj
}

export function addSecretIngredient(arr1, arr2) {
  arr2.push(...arr1.slice(-1))
}

export function scaleRecipe(recipe, numPortions) {
  const obj = {}
  for (let key in recipe) {
    obj[key] = recipe[key] * numPortions / 2
  }

  return obj
}
